consensus-types/hdiff/state_diff.go

```diff
@@ -0,0 +1,2101 @@
+	pendingPartialWithdrawalsDiffLength := int(binary.LittleEndian.Uint64((*data)[8:16])) // lint:ignore uintcast
+	if len(*data) < 16+pendingPartialWithdrawalsDiffLength*pendingPartialWithdrawalLength {
+		return errors.Wrap(errDataSmall, "pendingPartialWithdrawalsDiff")
+	}
+	ret.pendingPartialWithdrawalsDiff = make([]*ethpb.PendingPartialWithdrawal, pendingPartialWithdrawalsDiffLength)
+	cursor := 16
+	for i := range pendingPartialWithdrawalsDiffLength {
+		ret.pendingPartialWithdrawalsDiff[i] = &ethpb.PendingPartialWithdrawal{
+			Index:             primitives.ValidatorIndex(binary.LittleEndian.Uint64((*data)[cursor : cursor+8])),
+			Amount:            binary.LittleEndian.Uint64((*data)[cursor+8 : cursor+16]),
+			WithdrawableEpoch: primitives.Epoch(binary.LittleEndian.Uint64((*data)[cursor+16 : cursor+24])),
+		}
+		cursor += pendingPartialWithdrawalLength
+	}
+	*data = (*data)[cursor:]
+	return nil
+}
+
+func (ret *stateDiff) readPendingConsolidations(data *[]byte) error {
+	if len(*data) < 16 {
+		return errors.Wrap(errDataSmall, "pendingConsolidations")
+	}
+	ret.pendingConsolidationsIndex = binary.LittleEndian.Uint64((*data)[:8])
+	pendingConsolidationsDiffsLength := int(binary.LittleEndian.Uint64((*data)[8:16])) // lint:ignore uintcast
+	if len(*data) < 16+pendingConsolidationsDiffsLength*pendingConsolidationLength {
+		return errors.Wrap(errDataSmall, "pendingConsolidationsDiffs")
+	}
+	ret.pendingConsolidationsDiffs = make([]*ethpb.PendingConsolidation, pendingConsolidationsDiffsLength)
+	cursor := 16
+	for i := range pendingConsolidationsDiffsLength {
+		ret.pendingConsolidationsDiffs[i] = &ethpb.PendingConsolidation{
+			SourceIndex: primitives.ValidatorIndex(binary.LittleEndian.Uint64((*data)[cursor : cursor+8])),
+			TargetIndex: primitives.ValidatorIndex(binary.LittleEndian.Uint64((*data)[cursor+8 : cursor+16])),
+		}
+		cursor += pendingConsolidationLength
+	}
+	*data = (*data)[cursor:]
+	return nil
+}
+
+func (ret *stateDiff) readProposerLookahead(data *[]byte) error {
+	if len(*data) < proposerLookaheadLength {
+		return errors.Wrap(errDataSmall, "proposerLookahead data")
+	}
+	// Read the proposer lookahead (2 * SlotsPerEpoch uint64 values)
+	numProposers := 2 * fieldparams.SlotsPerEpoch
+	ret.proposerLookahead = make([]uint64, numProposers)
+	for i := 0; i < numProposers; i++ {
+		ret.proposerLookahead[i] = binary.LittleEndian.Uint64((*data)[i*8 : (i+1)*8])
+	}
+	*data = (*data)[proposerLookaheadLength:]
+	return nil
+}
+
+// newStateDiff deserializes a new stateDiff object from the given data.
+func newStateDiff(input []byte) (*stateDiff, error) {
+	data, err := snappy.Decode(nil, input)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to decode snappy")
+	}
+	ret := &stateDiff{}
+	if err := ret.readTargetVersion(&data); err != nil {
+		return nil, err
+	}
+	if err := ret.readSlot(&data); err != nil {
+		return nil, err
+	}
+	if err := ret.readFork(&data); err != nil {
```

Comment:
I feel like it would be very easy to forget something here, I wonder what would be the best way to keep track to make sure we add a new time every time we touch the state